### PR TITLE
Entitlements logging config options

### PR DESCRIPTION
**Pull request by mlw**
Adds two configuration options supporting entitlement logging:
* `EntitlementsTeamIDFilter` - Filter out entitlements from binaries from configured organizations
* `EntitlementsPrefixFilter` - Filter out specific entitlement prefixes

### CODE REVIEW

File: Source/common/SNTConfigurator.h [hunk 1]
```diff
@@ -642,6 +642,18 @@
 ///
 @property(readonly, nonatomic) NSUInteger metricExportTimeout;
 
+///
```
Comment (kallsyms) [inline Source/common/SNTConfigurator.h:645]: comments are flipped :)

File: Source/santad/Santad.mm [hunk 2]
```diff
@@ -358,6 +358,52 @@ void SantadMain(std::shared_ptr<EndpointSecurityAPI> esapi, std::shared_ptr<Logg
...
+
+              // Get the value from the configurator since that method ensures proper structure
+              [exec_controller
+                updateEntitlementsTeamIDFilter:[configurator entitlementsTeamIDFilter]];
+
+              // Clear the AuthResultCache, then clear the ES cache to ensure
+              // future execs get SNTCachedDecision entitlement values filtered
+              // with the new settings.
+              auth_result_cache->FlushCache(FlushCacheMode::kAllCaches,
+                                            FlushCacheReason::kEntitlementsPrefixFilterChanged);
```
Comment (kallsyms) [inline Source/santad/Santad.mm:381]:
```suggestion
                                            FlushCacheReason::kEntitlementsTeamIDFilterChanged);
```

File: Source/common/SNTConfigurator.h (hunk 1 above)
Comment (mlw) [reply Source/common/SNTConfigurator.h:645]: Oops

File: Source/santad/Santad.mm (hunk 2 above)
Comment (mlw) [reply Source/santad/Santad.mm]: Who knew there could be downsides to copy/paste

File: Source/santad/SNTExecutionController.mm [hunk 3]
```diff
@@ -38,18 +43,39 @@
...
+#include "absl/synchronization/mutex.h"
 
+using santa::common::PrefixTree;
+using santa::common::Unit;
 using santa::santad::TTYWriter;
 using santa::santad::event_providers::endpoint_security::Message;
 
 static const size_t kMaxAllowedPathLength = MAXPATHLEN - 1;  // -1 to account for null terminator
 
+void UpdateTeamIDFilterLocked(std::set<std::string> &filterSet, NSArray<NSString *> *filter) {
```
Comment (pmarkowsky) [inline Source/santad/SNTExecutionController.mm:62]: nit: We should probably write up somewhere that our "Locked" naming convention means we assume the lock is taken already before calling..

File: Source/santad/SNTExecutionController.mm [hunk 4]
```diff
@@ -63,6 +89,9 @@ @interface SNTExecutionController ()
 
 @implementation SNTExecutionController {
   std::shared_ptr<TTYWriter> _ttyWriter;
+  absl::Mutex _filterMutex;
```
Comment (pmarkowsky) [inline Source/santad/SNTExecutionController.mm:92]: We should probably name this `_entitlementFilterMutex`. While we're not filtering anything else today if we we'll need to deconflict the name later.

File: Source/santad/SNTPolicyProcessor.m [hunk 5]
```diff
@@ -94,8 +98,17 @@ - (nonnull SNTCachedDecision *)decisionForFileInfo:(nonnull SNTFileInfo *)fileIn
...
 
-      cd.entitlements =
-        [csInfo.signingInformation[(__bridge NSString *)kSecCodeInfoEntitlementsDict] sntDeepCopy];
+      NSDictionary *entitlements =
+        csInfo.signingInformation[(__bridge NSString *)kSecCodeInfoEntitlementsDict];
+
+      if (entitlementsFilterCallback) {
+        NSDictionary *filtered = entitlementsFilterCallback(entitlements);
+        cd.entitlements = filtered;
+        cd.entitlementsFiltered = filtered.count == entitlements.count;
```
Comment (pmarkowsky) [inline Source/santad/SNTPolicyProcessor.m:107]:
```suggestion
        // Set entitlements and filter status based on reduced number of entitlements.
        cd.entitlements = entitlementsFilterCallback(entitlements);
        cd.entitlementsFiltered = cd.entitlements.count == entitlements.count;
```

Seems like we don't need the filtered dictionary here.

**Review by pmarkowsky — CHANGES_REQUESTED** [review]
Only thing is renaming the mutex. Otherwise this seems fine. 

Please treat the other suggestions as nits.

**Review by pmarkowsky — APPROVED** [review]
LGTM. 

Thanks for making that change.

**Pull request merged**